### PR TITLE
Moving time increment higher up in the hierarchy 

### PR DESCRIPTION
**Pull request by shabnamjs**

### CODE REVIEW

**Comment by joshua-white** [comment]
@CusiniM and @rrsettgast, we are working on enabling the viscoplastic models.  To do so, we need to access the time increment `dt` inside the assembly kernels where the constitutive models will be called.  @shabnamjs is handling this by moving the location of the time increment member variable up further in the implicit finite element kernel hierarchy.  We wanted to run this approach by you before we get too far.  This will add another variable to the function signatures and take up extra register space (albeit just a scalar).

File: src/coreComponents/physicsSolvers/solidMechanics/kernels/ExplicitSmallStrain.hpp [hunk 1]
```diff
@@ -203,7 +203,7 @@ class ExplicitSmallStrain : public finiteElement::KernelBase< SUBREGION_TYPE,
   arrayView2d< real64, nodes::ACCELERATION_USD > const m_acc;
 
   /// The time increment for this time integration step.
-  real64 const m_dt;
+  real64 const m_dt; ///TODO: Consider moving to finite element kernel base?
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/solidMechanics/kernels/ExplicitSmallStrain.hpp:206]: yeah, I think you could probably just move it to the base class. At the end of the day your constitutive model will work with any time integration option.

**Review by CusiniM — APPROVED** [review]
I am okay with these changes. I don't really see a better way of doing this.

**Comment by shabnamjs** [comment]
@rrsettgast and @CusiniM , we think this PR is ready for your review, and it is passing all the checks (except for doxygen). I would appreciate it if you could take a look at this PR.

File: src/coreComponents/constitutive/solid/PorousSolid.cpp [hunk 2]
```diff
@@ -28,6 +28,7 @@
 #include "DamageSpectral.hpp"
 #include "DamageVolDev.hpp"
 #include "ModifiedCamClay.hpp"
+#include "DuvautLionsSolid.hpp"
 
 namespace geos
 {
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/solid/PorousSolid.cpp:34]: Porous Viscous models, such as `typedef PorousSolid< DuvautLionsSolid< DruckerPrager > > PorousViscoDruckerPrager` , should defined and registered.

File: src/coreComponents/constitutive/ConstitutivePassThru.hpp [hunk 3]
```diff
@@ -100,6 +100,9 @@ struct ConstitutivePassThru< SolidBase >
     ConstitutivePassThruHandler< DamageSpectral< ElasticIsotropic >,
                                  DamageVolDev< ElasticIsotropic >,
                                  Damage< ElasticIsotropic >,
+                                 DuvautLionsSolid< DruckerPrager >,
+                                 DuvautLionsSolid< DruckerPragerExtended >,
+                                 DuvautLionsSolid< ModifiedCamClay >,
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/ConstitutivePassThru.hpp:105]: In the this file,  Porous Viscous models should also be included in the `ConstitutivePassThruHandler` (line 235, 316 and 339)

**Review by jhuang2601 — APPROVED** [review]
Nice work! Thanks for enabling this modeling capability.

File: src/coreComponents/constitutive/solid/PorousSolid.cpp [hunk 4]
```diff
@@ -56,6 +57,10 @@ typedef PorousSolid< Damage< ElasticIsotropic > > PorousDamageElasticIsotropic;
 typedef PorousSolid< DamageSpectral< ElasticIsotropic > > PorousDamageSpectralElasticIsotropic;
 typedef PorousSolid< DamageVolDev< ElasticIsotropic > > PorousDamageVolDevElasticIsotropic;
 typedef PorousSolid< ModifiedCamClay > PorousModifiedCamClay;
+typedef PorousSolid< DuvautLionsSolid< ModifiedCamClay > > PorousViscoModifiedCamClay;
+typedef PorousSolid< DuvautLionsSolid< DruckerPrager > > PorousViscoDruckerPrager;
+typedef PorousSolid< DuvautLionsSolid< DruckerPragerExtended > > PorousViscoDruckerPragerExtended;
+
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/solid/PorousSolid.cpp:63]:
```suggestion
typedef PorousSolid< DuvautLionsSolid< DruckerPrager > > PorousViscoDruckerPrager;
typedef PorousSolid< DuvautLionsSolid< DruckerPragerExtended > > PorousViscoDruckerPragerExtended;
typedef PorousSolid< DuvautLionsSolid< ModifiedCamClay > > PorousViscoModifiedCamClay;
typedef PorousSolid< ModifiedCamClay > PorousModifiedCamClay;

```

File: src/coreComponents/constitutive/solid/PorousSolid.cpp [hunk 5]
```diff
@@ -67,6 +72,10 @@ REGISTER_CATALOG_ENTRY( ConstitutiveBase, PorousDamageElasticIsotropic, string c
 REGISTER_CATALOG_ENTRY( ConstitutiveBase, PorousDamageSpectralElasticIsotropic, string const &, Group * const )
 REGISTER_CATALOG_ENTRY( ConstitutiveBase, PorousDamageVolDevElasticIsotropic, string const &, Group * const )
 REGISTER_CATALOG_ENTRY( ConstitutiveBase, PorousModifiedCamClay, string const &, Group * const )
+REGISTER_CATALOG_ENTRY( ConstitutiveBase, PorousViscoModifiedCamClay, string const &, Group * const )
+REGISTER_CATALOG_ENTRY( ConstitutiveBase, PorousViscoDruckerPrager, string const &, Group * const )
+REGISTER_CATALOG_ENTRY( ConstitutiveBase, DruckerPragerExtended, string const &, Group * const )
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/solid/PorousSolid.cpp:77]:
```suggestion
REGISTER_CATALOG_ENTRY( ConstitutiveBase, PorousViscoDruckerPrager, string const &, Group * const )
REGISTER_CATALOG_ENTRY( ConstitutiveBase, PorousViscoDruckerPragerExtended, string const &, Group * const )
REGISTER_CATALOG_ENTRY( ConstitutiveBase, PorousViscoModifiedCamClay, string const &, Group * const )
```

**Review by jhuang2601 — APPROVED** [review]

**Comment by jhuang2601** [comment]
For a wellbore problem, I've compared a normal DruckerPrager model with a ViscoDruckerPrager model with ultra-low relaxation time and identical results can be obtained and matched with the corresponding analytical solution. 

For a stress relaxation case, GEOS results for ViscoDruckerPrager model correlates well with the semi-analytical solution.

Based on my current tests, this PR is ready to be merged and I'll create a separate PR for the analytical verification.

**Review by joshua-white — APPROVED** [review]

**Comment by joshua-white** [comment]
@shabnamjs, this looks good to go.  I've put you on the merge queue.  We will add the PoroViscous models in a separate PR, along with @jhuang2601 analytical verification tests.

**Comment by TotoGaz** [comment]
Hello, can you check that this does not require a rebaseline?

**Comment by shabnamjs** [comment]
> Hello, can you check that this does not require a rebaseline?

It passed the integrated tests and I did not modify any of the integrated tests, so I think it does not require a rebaseline.

**Comment by TotoGaz** [comment]
> It passed the integrated tests and I did not modify any of the integrated tests, so I think it does not require a rebaseline.

OK then!

**Comment by jhuang2601** [comment]
> Hello, can you check that this does not require a rebaseline?

Smoke test will be added in a separate PR (#2647) to protect this modeling capability.

**Pull request merged**